source/material_model/viscoelastic.cc

```diff
@@ -0,0 +1,700 @@
+
+          // Arithmetic averaging of thermal conductivities
+          // This may not be strictly the most reasonable thing, but for most Earth materials we hope
+          // that they do not vary so much that it is a big problem.
+          out.thermal_conductivities[i] = average_value(volume_fractions, thermal_conductivities, arithmetic);
+
+          double density = 0.0;
+          for (unsigned int j=0; j < volume_fractions.size(); ++j)
+            {
+              // not strictly correct if thermal expansivities are different, since we are interpreting
+              // these compositions as volume fractions, but the error introduced should not be too bad.
+              const double temperature_factor= (1.0 - thermal_expansivities[j] * (temperature - reference_T));
+              density += volume_fractions[j] * densities[j] * temperature_factor;
+            }
+          out.densities[i] = density;
+
+          out.thermal_expansion_coefficients[i] = average_value(volume_fractions, thermal_expansivities, arithmetic);
+
+          // Compressibility at the given positions.
+          // The compressibility is given as
+          // $\frac 1\rho \frac{\partial\rho}{\partial p}$.
+          // (here we use an incompressible medium)
+          out.compressibilities[i] = 0.0;
+          // Pressure derivative of entropy at the given positions.
+          out.entropy_derivative_pressure[i] = 0.0;
+          // Temperature derivative of entropy at the given positions.
+          out.entropy_derivative_temperature[i] = 0.0;
+          // Change in composition due to chemical reactions at the
+          // given positions. The term reaction_terms[i][c] is the
+          // change in compositional field c at point i.
+          for (unsigned int c=0; c<in.composition[i].size(); ++c)
+            out.reaction_terms[i][c] = 0.0;
+
+          // Average viscosity
+          const double average_viscosity = calculate_average_viscosity(composition,
+                                                                       viscosities,
+                                                                       viscosity_averaging);
+          // Average elastic shear modulus
+          const double average_elastic_shear_modulus =
+            calculate_average_elastic_shear_modulus(composition,
+                                                    elastic_shear_moduli,
+                                                    viscosity_averaging);
+
+          // Average viscoelastic (e.g., effective) viscosity (equation 28 in Moresi et al., 2003, J. Comp. Phys.)
+          const double average_viscoelastic_viscosity =
+            calculate_average_viscoelastic_viscosity(average_viscosity,
+                                                     average_elastic_shear_modulus,
+                                                     dte);
+
+          out.viscosities[i] = average_viscoelastic_viscosity;
```

Comment:
You can same some space by just assigning `out.viscosities[i] = calculate_average_viscoelastic_viscosity...`